src/include/miopen/graphapi/util.hpp

```diff
@@ -90,16 +91,23 @@ struct HeapPtrDeleter
     HeapPtrDeleter(const HeapPtrDeleter&) = delete;
     HeapPtrDeleter& operator=(const HeapPtrDeleter&) = delete;
 
+    friend void swap(HeapPtrDeleter& left, HeapPtrDeleter& right) noexcept
+    {
+        std::swap(left.mFn, right.mFn);
+    }
+
     HeapPtrDeleter(HeapPtrDeleter&& that) noexcept : mFn(std::move(that.mFn))
     {
         that.mFn = emptyFn;
     }
 
     HeapPtrDeleter& operator=(HeapPtrDeleter&& that) noexcept
     {
-        this->mFn(); // destruct self.
-        this->mFn = std::move(that.mFn);
-        that.mFn  = emptyFn;
+        if(this != &that)
+        {
+            HeapPtrDeleter tmp{std::move(that)};
+            swap(*this, tmp);
+        }
         return *this;
     }
```

Comment:
The problem is that the previous version is actually throw. As well as the current destructor:
```
~HeapPtrDeleter() { mFn(); }
```

> Invoking the target of an empty std::function results in `std::bad_function_call` exception being thrown.

`emptyFm` must not be used, I mean it can be, but it does not make much sense, since the check for `non-empty std::function` is required.

This particular code is fine now (with swap), but move constructor and destructor must be fixed as well as `emptyFn` must be removed.
You may keep it if you want, but even `const Fn` check for the function state in the destructor and thus generates unnecessary code - it's a tine overhead, but for no reason at all.